pkg/infrastructure/ibmcloud/clusterapi/clusterapi.go

```diff
@@ -232,3 +244,100 @@ func leftInContext(ctx context.Context) time.Duration {
+	cosInstanceName := ibmcloudic.COSInstanceName(in.InfraID)
+	var cosInstanceNotFoundError *ibmcloudic.COSResourceNotFoundError
+	cosInstance, err := client.GetCOSInstanceByName(ctx, cosInstanceName)
+	if err != nil {
+		if errors.As(err, &cosInstanceNotFoundError) {
+			// Attempt to create the COS Instance, since it was not found.
+			logrus.Debugf("creating cos instance: %s", cosInstanceName)
+			cosInstance, err = client.CreateCOSInstance(ctx, cosInstanceName, *resourceGroup.ID)
+			if err != nil {
+				return nil, fmt.Errorf("failed creating ignition COS instance: %w", err)
+			}
+			logrus.Debugf("created cos instance: %s", cosInstanceName)
+		} else {
+			return nil, fmt.Errorf("failed retrieving cos instance %s for ignition: %w", cosInstanceName, err)
+		}
+	}
+
+	// Create new bucket for bootstrap's temporary Ignition Config.
+	logrus.Debugf("fetching cos instance for cluster: %s", cosInstanceName)
+	bucketName := ibmcloudbootstrap.GetIgnitionBucketName((in.InfraID))
+	logrus.Debugf("creating cos bucket for bootstrap ignition config: %s", bucketName)
+	err = client.CreateCOSBucket(ctx, *cosInstance.ID, bucketName, region)
+	if err != nil {
+		return nil, fmt.Errorf("failed creating ignition COS bucket: %w", err)
+	}
+	logrus.Infof("created cos bucket for bootstrap ignition config: %s/%s", cosInstanceName, bucketName)
+
+	// Default to using the direct regional COS endpoint.
+	cosEndpoint := fmt.Sprintf("s3.direct.%s.cloud-object-storage.appdomain.cloud", region)
+	// Check whether an endpoint override was provided for COS.
+	if endpointURL := ibmcloudtypes.CheckServiceEndpointOverride(configv1.IBMCloudServiceCOS, in.InstallConfig.Config.IBMCloud.ServiceEndpoints); endpointURL != "" {
+		cosEndpoint = endpointURL
+	}
+
+	// Upload Ignition Config to COS bucket.
+	logrus.Debugf("uploading bootstrap ignition config to bucket: %s", bucketName)
+	ignitionFile := ibmcloudbootstrap.GetIgnitionFileName()
+	err = client.CreateCOSObject(ctx, in.BootstrapIgnData, ignitionFile, *cosInstance.ID, bucketName, region)
+	if err != nil {
+		return nil, fmt.Errorf("failed uploading ignition data: %w", err)
+	}
+	logrus.Debugf("bootstrap ignition config upload complete to %s/%s/%s", cosInstanceName, bucketName, ignitionFile)
+
+	ignitionURL := url.URL{
+		Scheme: "https",
+		Host:   cosEndpoint,
+		Path:   fmt.Sprintf("%s/%s", bucketName, ignitionFile),
+	}
+
+	// Build Ignition Config for Secret to direct bootstrap to consume COS Ignition Config.
+	logrus.Debugf("building ignition config data for bootstrap secret")
+	// Get IAM token for bootstrap node to access the Ignition config in COS.
+	iamToken, err := metadata.GetIAMToken(client.GetAPIKey())
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve iam token for ignition: %w", err)
+	}
+
+	logrus.Debugf("building bootstrap ignition config for bootstrap secret")
+	// NOTE(cjschaef): Replace the reliance on using the IAM token with a Service ID credential, when working with the COS Instance.
```

Comment:
Is this move to using Service ID credential happening in this PR or will this happen in a future update? Similar comment in GenerateIgnitionShimWithCredentials().

Reply:
This will happen in a future update. Those credentials need to be generated during Preprovision.